File: app/src/main/java/com/cwlarson/deviceid/database/Device.java
package com.cwlarson.deviceid.database;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.provider.Settings;
import android.telephony.TelephonyManager;
import android.util.Log;

import com.cwlarson.deviceid.R;
import com.cwlarson.deviceid.databinding.Item;
import com.cwlarson.deviceid.databinding.ItemType;
import com.cwlarson.deviceid.databinding.UnavailableItem;
import com.cwlarson.deviceid.databinding.UnavailablePermission;
import com.cwlarson.deviceid.databinding.UnavailableType;

class Device {
    private final String TAG = Device.class.getSimpleName();
    private final Context context;

    Device(Activity activity, AppDatabase db){
        this.context = activity.getApplicationContext();
        //Set Device Tiles
        ItemAdder itemAdder = new ItemAdder(context, db);
        itemAdder.addItems(getIMEI());
        itemAdder.addItems(getDeviceModel());
        itemAdder.addItems(getSerial());
        itemAdder.addItems(getAndroidID());
    }

    @SuppressLint({"HardwareIds", "MissingPermission"})
    private Item getIMEI() {
        Item item = new Item("IMEI / MEID",ItemType.DEVICE);
        try {
            // Request permission for IMEI/MEID for Android M+
            if (new Permissions(context).hasPermission(UnavailablePermission.MY_PERMISSIONS_REQUEST_READ_PHONE_STATE)) {
                TelephonyManager telephonyManager = (TelephonyManager) context.getSystemService(Context
                    .TELEPHONY_SERVICE);
                item.setSubtitle(telephonyManager.getDeviceId());
            } else {
                item.setUnavailableitem(new UnavailableItem(UnavailableType.NEEDS_PERMISSION,
                    context.getResources().getString(R.string.phone_permission_denied),
                    UnavailablePermission.MY_PERMISSIONS_REQUEST_READ_PHONE_STATE));
            }
        } catch (Exception e) {
            Log.w(TAG, "Null in getIMEI");
        }
        return item;
    }

    private Item getDeviceModel(){
        Item item = new Item("Model Number",ItemType.DEVICE);
        String device="",manufacturer,product,model;
        try {
            manufacturer = (Build.MANUFACTURER == null || Build.MANUFACTURER.length() == 0) ? "" : Build.MANUFACTURER;
            product = (Build.PRODUCT == null || Build.PRODUCT.length() == 0) ? "" : Build.PRODUCT;
            model = (Build.MODEL == null || Build.MODEL.length() == 0) ? "" : Build.MODEL;
            if (model.startsWith(manufacturer)) {
                device = model + " (" + product + ")";
            } else {
                device = manufacturer + " " + model + " (" + product + ")";
            }
        } catch (Exception e){
            Log.w(TAG, "Null in getDeviceModel");
        }
        item.setSubtitle(Character.isUpperCase(device.charAt(0)) ? device : Character.toUpperCase(device.charAt(0))+device.substring(1));
        return item;
    }

    @SuppressLint("HardwareIds")
    private Item getSerial() {
        Item item = new Item("Serial",ItemType.DEVICE);
        item.setSubtitle(Build.SERIAL);
        return item;
    }

    @SuppressLint("HardwareIds")
    private Item getAndroidID() {
        Item item = new Item("Android/Hardware ID",ItemType.DEVICE);
        try {
            item.setSubtitle(Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ANDROID_ID));
        } catch (Exception e){
            Log.w(TAG, "Null in getAndroidID");
        }
        return item;
    }
}

File: app/src/main/java/com/cwlarson/deviceid/database/Device.kt
package com.cwlarson.deviceid.database

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.net.Uri
import android.os.Build
import android.provider.Settings
import com.cwlarson.deviceid.R
import com.cwlarson.deviceid.databinding.*
import com.cwlarson.deviceid.util.hasPermission
import com.cwlarson.deviceid.util.telephonyManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import timber.log.Timber
import java.util.*

class Device(private val context: Context, db: AppDatabase, scope: CoroutineScope) {
    init {
        scope.launch {
            db.addItems(context, imei(), deviceModel(), serial(), androidID(), gsfid())
        }
    }

    // Request permission for IMEI/MEID for Android M+
    @SuppressLint("HardwareIds", "MissingPermission")
    private fun imei(): Item = Item("IMEI / MEID", ItemType.DEVICE).apply {
        if(Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            try {
                if (context.hasPermission(UnavailablePermission.MY_PERMISSIONS_REQUEST_READ_PHONE_STATE)) {
                    val telephonyManager = context.telephonyManager
                    subtitle = @Suppress("DEPRECATION") telephonyManager.deviceId
                } else
                        unavailableItem = UnavailableItem(UnavailableType.NEEDS_PERMISSION,
                                context.resources.getString(R.string.permission_item_subtitle,
                                        context.packageManager.getPermissionInfo(Manifest.permission.READ_PHONE_STATE, 0)
                                                .loadLabel(context.packageManager).toString()
                                                .toUpperCase(Locale.getDefault())),
                                UnavailablePermission.MY_PERMISSIONS_REQUEST_READ_PHONE_STATE)
            } catch (e: Exception) {
                Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
            }
        } else {
            unavailableItem = UnavailableItem(UnavailableType.NO_LONGER_POSSIBLE,
                    Build.VERSION_CODES.O.sdkToVersion())
        }
    }

    private fun deviceModel() = Item("Model Number", ItemType.DEVICE).apply {
        var device = ""
        val manufacturer: String
        val product: String
        val model: String
        try {
            manufacturer = if (Build.MANUFACTURER == null || Build.MANUFACTURER.isEmpty()) "" else Build.MANUFACTURER
            product = if (Build.PRODUCT == null || Build.PRODUCT.isEmpty()) "" else Build.PRODUCT
            model = if (Build.MODEL == null || Build.MODEL.isEmpty()) "" else Build.MODEL
            device = if (model.startsWith(manufacturer)) {
                "$model ($product)"
            } else {
                "$manufacturer $model ($product)"
            }.capitalize()
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
        subtitle = device
    }

    @SuppressLint("MissingPermission")
    private fun serial() = Item("Serial", ItemType.DEVICE).apply {
        if(Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            @Suppress("DEPRECATION") Build.SERIAL
        } else {
            unavailableItem = UnavailableItem(UnavailableType.NO_LONGER_POSSIBLE,
                    Build.VERSION_CODES.O.sdkToVersion())
        }
    }

    @SuppressLint("HardwareIds")
    private fun androidID() = Item("Android/Hardware ID", ItemType.DEVICE).apply {
        try {
            subtitle = Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun gsfid() = Item("Google Service Framework (GSF) ID", ItemType.DEVICE).apply {
        try {
            val uri = Uri.parse("content://com.google.android.gsf.gservices")
            val idKey = "android_id"
            val params = arrayOf(idKey)
            val c = context.contentResolver.query(uri, null, null, params, null)
            if (c != null && (!c.moveToFirst() || c.columnCount < 2)) {
                if (!c.isClosed) c.close()
            } else {
                try {
                    c?.let {
                        val result = java.lang.Long.toHexString(java.lang.Long.parseLong(it.getString(1)))
                        if (!it.isClosed) it.close()
                        subtitle = result
                    }
                } catch (e: Exception) {
                    if (c?.isClosed != true) c?.close()
                }

            }
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }
}
